storage/notifications/bridge.go

```diff
@@ -0,0 +1,139 @@
+package notifications
+
+import (
+	"time"
+
+	"github.com/docker/distribution/manifest"
+
+	"code.google.com/p/go-uuid/uuid"
+	"github.com/docker/distribution/digest"
+	"github.com/docker/distribution/storage"
+)
+
+type bridge struct {
+	ub     URLBuilder
+	actor  ActorRecord
+	source SourceRecord
+	sink   Sink
+}
+
+var _ Listener = &bridge{}
+
+// URLBuilder defines a subset of url builder to be used by the event listener.
+type URLBuilder interface {
+	BuildManifestURL(name, tag string) (string, error)
+	BuildBlobURL(name string, dgst digest.Digest) (string, error)
+}
+
+// NewBridge returns a notification listener that writes records to sink,
+// using the actor and source. Any urls populated in the events created by
+// this bridge will be created using the URLBuilder.
+func NewBridge(ub URLBuilder, source SourceRecord, actor ActorRecord, sink Sink) Listener {
+	return &bridge{
+		ub:     ub,
+		actor:  actor,
+		source: source,
+		sink:   sink,
+	}
+}
+
+func (b *bridge) ManifestPushed(repo storage.Repository, sm *manifest.SignedManifest) error {
+	return b.createManifestEventAndWrite(EventActionPush, repo, sm)
+}
+
+func (b *bridge) ManifestPulled(repo storage.Repository, sm *manifest.SignedManifest) error {
+	return b.createManifestEventAndWrite(EventActionPull, repo, sm)
+}
+
+func (b *bridge) ManifestDeleted(repo storage.Repository, sm *manifest.SignedManifest) error {
+	return b.createManifestEventAndWrite(EventActionDelete, repo, sm)
+}
+
+func (b *bridge) LayerPushed(repo storage.Repository, layer storage.Layer) error {
```

Comment:
For now, I want things to be consistent. There is a grip of renaming that needs to happen for that. The intended refactoring changes the name of `storage.Layer` to `storage.Blob`. We also need to move those definitions out of the storage package.

I'll probably file an issue with these plans later this week.